lib/charms/haproxy/v0/haproxy_route_tcp.py

```diff
@@ -0,0 +1,1738 @@
+    tls_terminate: bool = Field(description="Whether to enable tls termination.", default=True)
+
+    @model_validator(mode="after")
+    def assign_default_backend_port(self) -> "Self":
+        """Assign a default value to backend_port if not set.
+
+        The value is equal to the provider port.
+
+        Returns:
+            The model with backend_port default value applied.
+        """
+        if self.backend_port is None:
+            self.backend_port = self.port
+        return self
+
+    @model_validator(mode="after")
+    def sni_set_when_not_enforcing_tls(self) -> "Self":
+        """Check if sni is configured but TLS is disabled.
+
+        Raises:
+            ValueError: If sni is configured and TLS is disabled.
+
+        Returns:
+            The validated model.
+        """
+        if not self.enforce_tls and self.sni is not None:
+            raise ValueError("setting sni and disabling TLS are mutually exclusive.")
+        return self
+
+
+class HaproxyRouteTcpProviderAppData(_DatabagModel):
+    """haproxy-route provider databag schema.
+
+    Attributes:
+        endpoints: The list of proxied endpoints that maps to the backend.
+    """
+
+    endpoints: list[AnyUrl]
+
+
+class TcpRequirerUnitData(_DatabagModel):
+    """haproxy-route requirer unit data.
+
+    Attributes:
+        address: IP address of the unit.
+    """
+
+    address: IPvAnyAddress = Field(description="IP address of the unit.")
+
+
+@dataclass
+class HaproxyRouteTcpRequirerData:
+    """haproxy-route requirer data.
+
+    Attributes:
+        relation_id: Id of the relation.
+        application: Name of the requirer application.
+        application_data: Application data.
+        units_data: Units data
+    """
+
+    relation_id: int
+    application: str
+    application_data: TcpRequirerApplicationData
+    units_data: list[TcpRequirerUnitData]
+
+
+@dataclass
+class HaproxyRouteTcpRequirersData:
+    """haproxy-route requirers data.
+
+    Attributes:
+        requirers_data: List of requirer data.
+        relation_ids_with_invalid_data: List of relation ids that contains invalid data.
+    """
+
+    requirers_data: list[HaproxyRouteTcpRequirerData]
+    relation_ids_with_invalid_data: list[int]
+
+    @model_validator(mode="after")
+    def check_ports_unique(self) -> Self:
+        """Check that requirers define unique ports.
+
+        Returns:
+            The validated model, with invalid relation IDs updated in
+                `self.relation_ids_with_invalid_data`
+        """
+        # Maybe the logic here can be optimized, we want to keep track of
+        # the relation IDs that request overlapping ports to ignore them during
+        # rendering of the haproxy configuration.
+        port_to_relation_ids: dict[int, list[int]] = {}
+        for requirer_data in self.requirers_data:
+            if not port_to_relation_ids.get(requirer_data.application_data.port):
```

Comment:
You can use a defaultdict to simplify the logic

Reply:
updated, thanks!